README.md
# galois_field_2pm
A Rust library for representing and performing arithmetic of elements of Galois Fields of size 2<sup>M</sup>.

This library suppoorts addition, subtraction, multiplication, division, and inversion of elements in GF(2<sup>M</sup>) for 2 ≤ M ≤ 127.
Two implementations are supported. The implementations affect how multiplication, division, and inversion are computed.
The first implementation uses look up tables while the second implementation uses the Extended Euclidean Algorithm.

## Representing Galois Fields
GF(2<sup>M</sup>) is isomorphic to ${GF(2)[x] \over p(x)}$ where p(x) is an irreducible polynomial over GF(2) of degree M.
Elements of GF(2<sup>M</sup>) can therfore be uniquely be mapped to polynomials over GF(2) with degree less than M.
In order to represent all the polynomials over GF(2) with degree less than M, M bits are needed.

## Implementations
The look up table implementation can only be used when p(x) is a primitive polynomial with degree less than or equal to 16.
The other implementation will work with any irreducible polynomial up to degree 127.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2022-12-26
### Added
- Initial implementation


src/gf2.rs
use core::fmt;
use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};
use paste::paste;

mod clmul;
mod gf2_poly_div;

use crate::GaloisField;
use clmul::CarryLessMultiply;
use gf2_poly_div::GF2PolyDiv;

macro_rules! assign_operator_impl {
    ($($type:ty: $trait_name:ident: $trait_fn:ident: $op:tt,)*) => {
    $(
        paste! {
            impl<const POLY: u128> $trait_name for [<GF $type>]<POLY> {
                fn $trait_fn(&mut self, other: Self) {
                    *self = *self $op other;
                }
            }
        }
    )*
    }
}

macro_rules! setup_gf {
    ($($type:ty,)*) => {
    $(
        paste! {
            // Define the struct
            #[repr(transparent)]
            #[derive(Clone, Copy, Eq, PartialEq)]
            pub struct [<GF $type>]<const POLY: u128> {
                pub value: $type,
            }

            // Implement the traits
            impl<const POLY: u128> GaloisField for [<GF $type>]<POLY> {
                type StorageType = $type;

                const M: u128 = crate::calc_degree(POLY) as u128;
                const NUM_ELEM: u128 = 1 << Self::M;

                const ZERO: Self = Self {value: 0};
                const ONE: Self = Self {value: 1};

                fn inverse(&self) -> Self {
                    if *self == Self::ZERO {
                        panic!("Cannot take inverse of zero");
                    }
                    if *self == Self::ONE {
                        return Self::ONE;
                    }

                    let mut temp_t: Self::StorageType;
                    let mut t: Self::StorageType = 0;
                    let mut new_t: Self::StorageType = 1;

                    let mut quo = Self::ZERO;
                    let mut nt = Self::ZERO;

                    let mut new_remainder: Self::StorageType;
                    let mut remainder = self.value;
                    (quo.value, new_remainder) = $type::gf2_poly_div_poly(POLY, self.value);

                    nt.value = new_t;
                    temp_t = t;
                    t = new_t;
                    new_t = temp_t ^ ((quo * nt).value);

                    while new_remainder != 0 {
                        let temp_r = remainder;
                        remainder = new_remainder;
                        (quo.value, new_remainder) = $type::gf2_poly_div(temp_r, new_remainder);

                        nt.value = new_t;
                        temp_t = t;
                        t = new_t;
                        new_t = temp_t ^ ((quo * nt).value);
                    }

                    Self {
                        value: t
                    }
                }

                fn new(value: $type) -> Self {
                    Self {value: value}
                }

                fn validate(&self) -> bool {
                    (self.value as u128) >= Self::NUM_ELEM
                }
            }

            impl<const POLY: u128> fmt::Debug for [<GF $type>]<POLY> {
                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                    write!(f, "GF<{:#0X}>(value: 0x{:0width$x})", POLY, self.value, width = (Self::M as usize / 4))
                }
            }

            impl<const POLY: u128> fmt::Display for [<GF $type>]<POLY> {
                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                    write!(f, "0x{:0width$X}", self.value, width = (Self::M as usize / 4))
                }
            }

            impl<const POLY: u128> Add<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn add(self, other: Self) -> Self {
                    Self {
                        value: self.value ^ other.value,
                    }
                }
            }

            impl<const POLY: u128> Sub<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn sub(self, other: Self) -> Self {
                    Self {
                        value: self.value ^ other.value,
                    }
                }
            }

            impl<const POLY: u128> Mul<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn mul(self, other: Self) -> Self {
                    let hi = self.value.clmul_high(other.value);
                    let lo = self.value.clmul_low(other.value);

                    Self {
                        value: $type::gf2_poly_mod(hi, lo, POLY),
                    }
                }
            }

            impl<const POLY: u128> Div<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn div(self, other: Self) -> Self {
                    self * other.inverse()
                }
            }

            assign_operator_impl! {
                $type: AddAssign: add_assign: +,
                $type: SubAssign: sub_assign: -,
                $type: MulAssign: mul_assign: *,
                $type: DivAssign: div_assign: /,
            }
        }
    )*
    };
}

setup_gf! {
    u8,
    u16,
    u32,
    u64,
    u128,
}


src/gf2_lut.rs
use core::fmt;
use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};
use paste::paste;

use crate::GaloisField;

/// A trait used to indicate that the implementation of the Galois Field uses a look up table (LUT)
pub trait GaloisFieldLut: GaloisField {
    /// The primitive element of the field α. Stored as the value 2.
    const ALPHA: Self;

    /// Returns α<sup>power</sup>
    fn alpha_pow(power: isize) -> Self;

    /// For input α<sup>power</sup> returns power. For 0 returns -1.
    fn log_alpha(&self) -> isize;
}

macro_rules! assign_operator_impl {
    ($($type:ty: $trait_name:ident: $trait_fn:ident: $op:tt,)*) => {
    $(
        paste! {
            impl<const POLY: u128> $trait_name for [<GF $type>]<POLY> {
                fn $trait_fn(&mut self, other: Self) {
                    *self = *self $op other;
                }
            }
        }
    )*
    }
}

macro_rules! setup_gf {
    ($($type:ty,)*) => {
    $(
        paste! {
            // Define the struct
            #[repr(transparent)]
            #[derive(Clone, Copy, Eq, PartialEq)]
            pub struct [<GF $type>]<const POLY: u128> {
                pub value: $type,
            }

            // Implement the traits
            impl<const POLY: u128> GaloisField for [<GF $type>]<POLY> {
                type StorageType = $type;

                const M: u128 = crate::calc_degree(POLY) as u128;
                const NUM_ELEM: u128 = 1 << Self::M;

                const ZERO: Self = Self {value: 0};
                const ONE: Self = Self {value: 1};

                fn inverse(&self) -> Self {
                    if *self == Self::ZERO {
                        panic!("Can not take inverse of zero");
                    } else {
                        return Self::alpha_pow(Self::DEGREE_MOD - self.log_alpha());
                    }
                }

                fn new(value: $type) -> Self {
                    Self {value: value}
                }

                fn validate(&self) -> bool {
                    (self.value as u128) >= Self::NUM_ELEM
                }
            }

            impl<const POLY: u128> GaloisFieldLut for [<GF $type>]<POLY> {
                const ALPHA: Self = Self {value: 2};

                fn alpha_pow(power: isize) -> Self {
                    let mut pow = power % Self::DEGREE_MOD;
                    pow += Self::DEGREE_MOD;
                    pow %= Self::DEGREE_MOD;
                    Self { value: Self::TABLES.exp_tbl[pow as usize] }
                }

                fn log_alpha(&self) -> isize {
                    Self::TABLES.log_tbl[self.value as usize]
                }
            }

            // Implement all the behind the scenes detail
            struct [<Tables $type:upper>] {
                exp_tbl: [$type; 1 << $type::BITS],
                log_tbl: [isize; 1 << $type::BITS],
            }

            const fn [<generate_lut_ $type:lower>](poly: u128) -> [<Tables $type:upper>] {
                let m: usize = crate::calc_degree(poly) as usize;
                let num_elems: usize = (1 << m) - 1;
                let feedback_mask: $type = 1 << (m - 1);

                let mut exp_tbl: [$type; 1 << $type::BITS] = [0; 1 << $type::BITS];
                let mut log_tbl: [isize; 1 << $type::BITS] = [0; 1 << $type::BITS];

                log_tbl[0] = -1;
                let mut value: $type = 1;
                let mut i: usize = 0;
                while i < num_elems {
                    exp_tbl[i] = value;
                    log_tbl[exp_tbl[i] as usize] = i as isize;

                    let leading_1: bool = value >= feedback_mask;
                    value <<= 1;
                    if leading_1 {
                        value ^= poly as $type;
                    }
                    i += 1;
                }

                [<Tables $type:upper>] {
                    exp_tbl: exp_tbl,
                    log_tbl: log_tbl,
                }
            }

            impl<const POLY: u128> [<GF $type>]<POLY> {
                const TABLES : [<Tables $type:upper>] = [<generate_lut_ $type:lower>](POLY);
                const DEGREE_MOD: isize = (Self::NUM_ELEM as isize) - 1;
            }

            impl<const POLY: u128> fmt::Debug for [<GF $type>]<POLY> {
                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                    write!(f, "GF<{:#0X}>(value: 0x{:0width$x})", POLY, self.value, width = (Self::M as usize / 4))
                }
            }

            impl<const POLY: u128> fmt::Display for [<GF $type>]<POLY> {
                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                    write!(f, "0x{:0width$X}", self.value, width = (Self::M as usize / 4))
                }
            }

            impl<const POLY: u128> Add<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn add(self, other: Self) -> Self {
                    Self {
                        value: self.value ^ other.value,
                    }
                }
            }

            impl<const POLY: u128> Sub<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn sub(self, other: Self) -> Self {
                    Self {
                        value: self.value ^ other.value,
                    }
                }
            }

            impl<const POLY: u128> Mul<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn mul(self, other: Self) -> Self {
                    if (self == Self::ZERO) || (other == Self::ZERO) {
                        return Self::ZERO;
                    } else {
                        return Self::alpha_pow(self.log_alpha() + other.log_alpha());
                    }
                }
            }

            impl<const POLY: u128> Div<[<GF $type>]<POLY>> for [<GF $type>]<POLY> {
                type Output = Self;

                fn div(self, other: Self) -> Self {
                    if other == Self::ZERO {
                        panic!("Divide by 0");
                    } else if self == Self::ZERO {
                        return Self::ZERO;
                    } else {
                        return Self::alpha_pow(self.log_alpha() - other.log_alpha());
                    }
                }
            }

            assign_operator_impl! {
                $type: AddAssign: add_assign: +,
                $type: SubAssign: sub_assign: -,
                $type: MulAssign: mul_assign: *,
                $type: DivAssign: div_assign: /,
            }
        }
    )*
    }
}

setup_gf! {
    u8,
    u16,
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! lut_specific_tests {
        ($($type:ty: $poly:expr,)*) => {
        $(
            paste! {
                #[test]
                fn [<alpha_pow_log_alpha_ $poly>]() {
                    type GF = [<GF $type>]<$poly>;
                    assert_eq!(GF::ZERO.log_alpha(), -1);

                    for i in 1..GF::DEGREE_MOD {
                        assert_eq!(GF::alpha_pow(i as isize).log_alpha(), i as isize);
                    }
                }
            }
        )*
        }
    }

    lut_specific_tests! {
        u16: 0x3,
        u16: 0x7,
        u16: 0xb,
        u16: 0x13,
        u8: 0x25,
        u8: 0x43,
        u8: 0x83,
        u8: 0x11d,
    }
}


src/gf2/clmul.rs
pub trait CarryLessMultiply {
    fn clmul_low(&self, rhs: Self) -> Self;
    fn clmul_high(&self, rhs: Self) -> Self;
}

macro_rules! clmul_impl {
    ($($type:ty,)*) => {
    $(
        impl CarryLessMultiply for $type {
            fn clmul_low(&self, rhs: Self) -> Self {
                const NUM_BITS: u32 = <$type>::BITS;
                let mut output: Self = 0;

                for i in 0..NUM_BITS {
                    if ((rhs >> i) & 0x1) > 0 {
                        output ^= self << i;
                    }
                }

                output
            }

            fn clmul_high(&self, rhs: Self) -> Self {
                const NUM_BITS: u32 = <$type>::BITS;
                let mut output: Self = 0;

                for i in 1..NUM_BITS {
                    if ((rhs >> i) & 0x1) > 0 {
                        output ^= self >> (NUM_BITS - i);
                    }
                }

                output
            }
        }
    )*
    };
}

clmul_impl! {
    u8,
    u16,
    u32,
    u64,
    u128,
}


src/gf2/gf2_poly_div.rs
pub trait GF2PolyDiv {
    type Elem;

    // This is used for inverting an element in GF(2^M)
    fn gf2_poly_div(dividend: Self::Elem, divisor: Self::Elem) -> (Self::Elem, Self::Elem);

    // This is used for inverting an element in GF(2^M). Used to divide POLY by lower degree value.
    // Divisor must be > 1
    fn gf2_poly_div_poly(dividend: u128, divisor: Self::Elem) -> (Self::Elem, Self::Elem);

    // Take the modulus of (hi | lo) by POLY
    fn gf2_poly_mod(hi: Self::Elem, lo: Self::Elem, poly: u128) -> Self::Elem;
}

macro_rules! gf2_poly_div_impl {
    ($($type:ty,)*) => {
    $(
        impl GF2PolyDiv for $type {
            type Elem = $type;

            fn gf2_poly_div(dividend: Self::Elem, divisor: Self::Elem) -> (Self::Elem, Self::Elem) {
                if divisor == 1 {
                    return (dividend, 0);
                }

                let mut quotient: Self::Elem = 0;
                let mut remainder: Self::Elem = 0;

                let deg_divisor = crate::calc_degree(divisor as u128);
                let d = deg_divisor - 1;
                let mask: Self::Elem = divisor & !(1 << deg_divisor);

                for cur_deg in (0..=crate::calc_degree(dividend as u128)).rev() {
                    let bit: bool = ((remainder >> d) & 0x1) > 0;
                    if bit {
                        quotient = (quotient << 1) | 0x1;
                    } else {
                        quotient <<= 1;
                    }

                    remainder = (remainder << 1) | ((dividend >> cur_deg) & 0x1);
                    if bit {
                        remainder ^= mask;
                    }
                }

                let bit_limit: Self::Elem = (1 << deg_divisor) - 1;
                remainder &= bit_limit;
                (quotient, remainder)
            }

            fn gf2_poly_div_poly(dividend: u128, divisor: Self::Elem) -> (Self::Elem, Self::Elem) {
                let mut quotient: Self::Elem = 0;
                let mut remainder: Self::Elem = 0;

                let deg_divisor = crate::calc_degree(divisor as u128);
                let d = deg_divisor - 1;
                let mask: Self::Elem = divisor & !(1 << deg_divisor);

                for cur_deg in (0..=crate::calc_degree(dividend as u128)).rev() {
                    let bit: bool = ((remainder >> d) & 0x1) > 0;
                    if bit {
                        quotient = (quotient << 1) | 0x1;
                    } else {
                        quotient <<= 1;
                    }

                    remainder = (remainder << 1) | (((dividend >> cur_deg) as Self::Elem) & 0x1) ;
                    if bit {
                        remainder ^= mask;
                    }
                }

                let bit_limit: Self::Elem = (1 << deg_divisor) - 1;
                remainder &= bit_limit;
                (quotient, remainder)
            }

            fn gf2_poly_mod(hi: Self::Elem, lo: Self::Elem, poly: u128) -> Self::Elem {
                let deg_divisor: usize = crate::calc_degree(poly) as usize;
                let d = deg_divisor - 1;
                let mask: Self::Elem = (poly & !(1u128 << deg_divisor)) as Self::Elem;

                let mut remainder: Self::Elem = 0;

                let hi_deg = crate::calc_degree(hi as u128);
                let lo_deg = if hi_deg < 0 { crate::calc_degree(lo as u128) } else { (Self::Elem::BITS - 1) as i16};

                // Start with the hi portion
                for cur_deg in (0..=hi_deg).rev() {
                    let bit: bool = ((remainder >> d) & 0x1) > 0;
                    remainder = (remainder << 1) | ((hi >> cur_deg) & 0x1);
                    if bit {
                        remainder ^= mask;
                    }
                }

                // Now use the lo portion
                for cur_deg in (0..=lo_deg).rev() {
                    let bit: bool = ((remainder >> d) & 0x1) > 0;
                    remainder = (remainder << 1) | ((lo >> cur_deg) & 0x1);
                    if bit {
                        remainder ^= mask;
                    }
                }

                let bit_limit: Self::Elem = ((1u128 << deg_divisor) - 1) as Self::Elem;
                remainder &= bit_limit;

                remainder
            }
        }
    )*
    };
}

gf2_poly_div_impl! {
    u8,
    u16,
    u32,
    u64,
    u128,
}


src/lib.rs
//! Galois Field (2<sup>M</sup>) arithmetic support.
//!
//! galois_field_2pm provides wrappers for the u8, u16, u32, u64, and u128 types that implement Galois Field arithmetic. There are two implementations:
//! - A look up table based implementation for multiplication and division for Galois Fields defined by a primitive polynomial
//! - A computation based implementation for multiplication and division for Galois Fields defined by an irreducible polynomial
//!
//! # Getting Started
//!
//! To start using a Galois Field defined by the irreducible polynomial p(x) ∈ GF(2)\[x\] start by representing p(x) as a u128.
//!
//! An irrecuible polynomial of degree M defines the field GF(2<sup>M</sup>) = GF(2)\[x\] / p(x)
//!
//! Next we need to choose how the field is implemented:
//!
//!   - If p(x) is primitive and M ≤ 16 then the look up table implementation can be used (module gf2_lut)
//!
//!   - Else use the computation based implementation (module gf2)
//!
//! Lastly we must use one of the structs to represent the elements in the field. The struct GFuX can be used for M ≤ X.
//!
//! At this point we can define the GF type and use it.
//!
//! Consider the example below
//! ```
//! use galois_field_2pm::{GaloisField, gf2_lut};
//!
//! // First lets define out Galois Field
//! // For the irreducible polynomial p(x) = x^3 + x + 1 the u128 representation is given by 0xB.
//! // Note that this p(x) is primitive so we can use either gf2_lut of gf2
//! // Note that GFu8 can represent the field generate by 0xB as the degree of 0xB is 3 and 3 ≤ 8.
//! //     For p(x) = 0x211 (degree 9), GFu8 can not be used. GFu16 must be used instead
//! type GF = gf2_lut::GFu8::<0xB>; // Or alternatively type GF = gf2::GFu8::<0xB>;
//! let a = GF::ONE;
//! let b = GF::new(2);
//! let c = GF {value: 3};
//!
//! let d = (a + b * c).inverse();
//! ```

#[cfg(test)]
use paste::paste;

#[cfg(test)]
use rand::Rng;

use core::fmt::{Debug, Display};
use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};

pub mod gf2;
pub mod gf2_lut;

/// A trait used to indicate that a type can be used to represent the elements of a Galois Field.
pub trait GaloisField:
    Clone
    + Copy
    + Eq
    + PartialEq
    + Debug
    + Display
    + Add
    + AddAssign
    + Sub
    + SubAssign
    + Mul
    + MulAssign
    + Div
    + DivAssign
{
    /// The underlying type used to store the representation of an element in the field
    type StorageType;

    /// The degree of the polynomial used to define the field. Specifies GF(2<sup>M</sup>)
    const M: u128;

    /// The number of elements in the field. This value is 2<sup>M</sup>
    const NUM_ELEM: u128;

    /// The additive identity of the field
    const ZERO: Self;

    /// The multiplicative identity of the field
    const ONE: Self;

    /// Takes the inverse of an element in the field. Panics when inverting the additive identity (aka zero)
    fn inverse(&self) -> Self;

    /// Constructs a GF element using the underlying storage type
    fn new(value: Self::StorageType) -> Self;

    /// Used to check if the value stored is a valid element in the current field
    fn validate(&self) -> bool;
}

#[allow(dead_code)]
const fn calc_degree(x: u128) -> i16 {
    127 - (x.leading_zeros() as i16)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn calc_degree_test() {
        assert_eq!(calc_degree(0), -1);
        for i in 0..=127 {
            assert_eq!(calc_degree(1u128 << i), i);
        }
    }

    macro_rules! associative_test {
        ($type:ty, $op:tt) => {
            for i in 0..GF::NUM_ELEM {
                for j in 0..GF::NUM_ELEM {
                    for k in 0..GF::NUM_ELEM {
                        let a = GF::new(i as $type);
                        let b = GF::new(j as $type);
                        let c = GF::new(k as $type);
                        assert_eq!(a $op (b $op c), (a $op b) $op c);
                    }
                }
            }
        }
    }

    macro_rules! commutative_test {
        ($type:ty, $op:tt) => {
            for i in 0..GF::NUM_ELEM {
                for j in 0..GF::NUM_ELEM {
                    let a = GF::new(i as $type);
                    let b = GF::new(j as $type);
                    assert_eq!(a $op b, b $op a);
                }
            }
        }
    }

    macro_rules! identity_test {
        ($type:ty, $op:tt, $identity:tt) => {
            for i in 0..GF::NUM_ELEM {
                let a = GF::new(i as $type);
                assert_eq!(a $op GF::$identity, a);
            }
        }
    }

    macro_rules! inverse_addition_test {
        ($type:ty) => {
            for i in 0..GF::NUM_ELEM {
                let a = GF::new(i as $type);
                assert_eq!(a + a, GF::ZERO);
            }
        };
    }

    macro_rules! inverse_multiplication_test {
        ($type:ty) => {
            for i in 1..GF::NUM_ELEM {
                let a = GF::new(i as $type);
                assert_eq!(a * a.inverse(), GF::ONE);
                assert_eq!(a / a, GF::ONE);
            }
        };
    }

    macro_rules! addition_test {
        ($type:ty) => {
            associative_test!($type, +);
            commutative_test!($type, +);
            identity_test!($type, +, ZERO);
            inverse_addition_test!($type);
        }
    }

    macro_rules! multiplication_test {
        ($type:ty) => {
            associative_test!($type, *);
            commutative_test!($type, *);
            identity_test!($type, *, ONE);
            inverse_multiplication_test!($type);
        }
    }

    macro_rules! distributive_test {
        ($type:ty) => {
            for i in 0..GF::NUM_ELEM {
                for j in 0..GF::NUM_ELEM {
                    for k in 0..GF::NUM_ELEM {
                        let a = GF::new(i as $type);
                        let b = GF::new(j as $type);
                        let c = GF::new(k as $type);
                        assert_eq!(a * (b + c), a * b + a * c);
                    }
                }
            }
        };
    }

    macro_rules! field_test {
        ($($mod:tt: $type:ty: $poly:expr,)*) => {
        $(
            paste! {
                #[test]
                fn [<$mod _gf_ $poly>]() {
                    type GF = $mod::[<GF $type>]<$poly>;
                    addition_test!($type);
                    multiplication_test!($type);
                    distributive_test!($type);
                }

                #[test]
                #[should_panic]
                fn [<$mod _div_by_0_ $poly>]() {
                    type GF = $mod::[<GF $type>]<$poly>;
                    let _ = GF::ONE / GF::ZERO;
                }

                #[test]
                #[should_panic]
                fn [<$mod _inverse_of_0_ $poly>]() {
                    type GF = $mod::[<GF $type>]<$poly>;
                    let _ = GF::ZERO.inverse();
                }
            }
        )*
        }
    }

    field_test! {
        gf2_lut: u16: 0x3,
        gf2_lut: u16: 0x7,
        gf2_lut: u16: 0xb,
        gf2_lut: u16: 0x13,
        gf2_lut: u8: 0x25,
        gf2_lut: u8: 0x43,
        gf2_lut: u8: 0x83,
        gf2_lut: u8: 0x11d,

        gf2: u16: 0x3,
        gf2: u16: 0x7,
        gf2: u16: 0xb,
        gf2: u16: 0x13,
        gf2: u8: 0x25,
        gf2: u8: 0x43,
        gf2: u8: 0x83,
        gf2: u8: 0x11d,
    }

    macro_rules! associative_spot_test {
        ($type:ty, $a:tt, $b:tt, $c:tt, $op:tt) => {
            for i in 0..NUM_VALS {
                for j in 0..NUM_VALS {
                    for k in 0..NUM_VALS {
                        assert_eq!($a[i] $op ($b[j] $op $c[k]), ($a[i] $op $b[j]) $op $c[k]);
                    }
                }
            }
        }
    }

    macro_rules! commutative_spot_test {
        ($type:ty, $a:tt, $b:tt, $op:tt) => {
            for i in 0..NUM_VALS {
                for j in 0..NUM_VALS {
                    assert_eq!($a[i] $op $b[j], $b[j] $op $a[i]);
                }
            }
        }
    }

    macro_rules! identity_spot_test {
        ($type:ty, $a:tt, $op:tt, $identity:tt) => {
            for i in 0..NUM_VALS {
                assert_eq!($a[i] $op GF::$identity, $a[i]);
            }
        }
    }

    macro_rules! inverse_addition_spot_test {
        ($type:ty, $a:tt) => {
            for i in 0..NUM_VALS {
                assert_eq!($a[i] + $a[i], GF::ZERO);
            }
        };
    }

    macro_rules! inverse_multiplication_spot_test {
        ($type:ty, $a:tt) => {
            for i in 0..NUM_VALS {
                assert_eq!($a[i] * $a[i].inverse(), GF::ONE);
                assert_eq!($a[i] / $a[i], GF::ONE);
            }
        };
    }

    macro_rules! addition_spot_test {
        ($type:ty, $a:tt, $b:tt, $c:tt) => {
            associative_spot_test!($type, $a, $b, $c, +);
            commutative_spot_test!($type, $a, $b, +);
            identity_spot_test!($type, $a, +, ZERO);
            inverse_addition_spot_test!($type, $a);
        }
    }

    macro_rules! multiplication_spot_test {
        ($type:ty, $a:tt, $b:tt, $c:tt) => {
            associative_spot_test!($type, $a, $b, $c, *);
            commutative_spot_test!($type, $a, $b, *);
            identity_spot_test!($type, $a, *, ONE);
            inverse_multiplication_spot_test!($type, $a);
        }
    }

    macro_rules! distributive_spot_test {
        ($type:ty, $a:tt, $b:tt, $c:tt) => {
            for i in 0..NUM_VALS {
                for j in 0..NUM_VALS {
                    for k in 0..NUM_VALS {
                        assert_eq!($a[i] * ($b[j] + $c[k]), $a[i] * $b[j] + $a[i] * $c[k]);
                    }
                }
            }
        };
    }

    macro_rules! field_spot_test {
        ($($mod:tt: $type:ty: $poly:expr,)*) => {
        $(
            paste! {
                #[test]
                fn [<$mod _spot_gf_ $poly>]() {
                    type GF = $mod::[<GF $type>]<$poly>;

                    const NUM_VALS: usize = 30;
                    let mut a: [GF; NUM_VALS] = [GF::ZERO; NUM_VALS];
                    let mut b: [GF; NUM_VALS] = [GF::ZERO; NUM_VALS];
                    let mut c: [GF; NUM_VALS] = [GF::ZERO; NUM_VALS];
                    for i in 0..NUM_VALS {
                        a[i] = GF::new(rand::thread_rng().gen_range(1..GF::NUM_ELEM) as $type);
                        b[i] = GF::new(rand::thread_rng().gen_range(1..GF::NUM_ELEM) as $type);
                        c[i] = GF::new(rand::thread_rng().gen_range(1..GF::NUM_ELEM) as $type);
                    }

                    addition_spot_test!($type, a, b, c);
                    multiplication_spot_test!($type, a, b, c);
                    distributive_spot_test!($type, a, b, c);
                }
            }
        )*
        }
    }

    field_spot_test! {
        gf2_lut: u16: 0x211,
        gf2_lut: u16: 0x409,
        gf2_lut: u16: 0x805,
        gf2_lut: u16: 0x1053,

        gf2: u8: 0x11b,
        gf2: u8: 0x1dd,
        gf2: u16: 0x211,
        gf2: u16: 0x409,
        gf2: u16: 0x805,
        gf2: u16: 0x1053,
        gf2: u32: 0x2_0009,
        gf2: u64: 0x2_0000_2001,
        gf2: u128: 0x2_0000_0000_0004_0001,
        gf2: u128: 0x8000_0000_0000_0000_0000_0000_0000_0003,
    }
}
